libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteRequest.kt

```diff
@@ -33,51 +33,29 @@ interface CarRouteRequestCallback {
  * Service class that requests routes for the preview screen.
  */
 class CarRouteRequest(
-    val mapboxNavigation: MapboxNavigation,
     private val routeOptionsInterceptor: CarRouteOptionsInterceptor,
-    private val navigationLocationProvider: NavigationLocationProvider,
-) {
-    internal var currentRequestId: Long? = null
-
-    suspend fun requestSync(placeRecord: PlaceRecord): List<NavigationRoute>? {
-        return suspendCancellableCoroutine { continuation ->
-            continuation.invokeOnCancellation { cancelRequest() }
-            request(
-                placeRecord,
-                object : CarRouteRequestCallback {
-
-                    override fun onRoutesReady(
-                        placeRecord: PlaceRecord,
-                        routes: List<NavigationRoute>
-                    ) {
-                        continuation.resume(routes)
-                    }
-
-                    override fun onUnknownCurrentLocation() {
-                        continuation.resume(value = null)
-                    }
-
-                    override fun onDestinationLocationUnknown() {
-                        continuation.resume(value = null)
-                    }
-
-                    override fun onNoRoutesFound() {
-                        continuation.resume(value = null)
-                    }
-                }
-            )
-        }
+) : MapboxNavigationObserver {
+    private var currentRequestId: Long? = null
+    private var mapboxNavigation: MapboxNavigation? = null
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        this.mapboxNavigation = mapboxNavigation
+    }
+
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        cancelRequest()
+        this.mapboxNavigation = null
     }
 
     /**
      * When a search result was selected, request a route.
-     *
-     * @param searchResults potential destinations for directions
      */
     fun request(placeRecord: PlaceRecord, callback: CarRouteRequestCallback) {
-        currentRequestId?.let { mapboxNavigation.cancelRouteRequest(it) }
+        val mapboxNavigation = this.mapboxNavigation ?: return
```

Comment:
Should we use !! here?

Reply:
I thought about this too much, but also established an opinion. I agree with you that a no-op is not great here because `CarRouteRequestCallback` is not called and there is no signal that the request did not happen.

Considering that this is just one new type of many possible failures. I think the error should be sent to `CarRouteRequestCallback`. So for now it's sending it to `onNoRoutesFound`

Reply:
Arguable that this is a developer error so it should crash. But it is also kind of nitpicking to crash in this case because you may want to request routes from things that are not lifecycle aware. 

The `CarRouteRequest` is lifecycle aware so the caller does not have to be 🤔 

``` kotlin
object SomeUtility {
--snip--
  fun goToNextDestination(placeRecord) {
    MapboxNavigationApp.getObserver(CarRouteRequest::class).request(placeRecord, callback)
  }
}
```

Reply:
And then the reason why it's good to always have a callback respond. Is because otherwise people will build "infinite spinners"

``` kotlin
// simplified version of the callback to illustrate the point.
// stopLoading would not be called if the `CarRouteRequest` is detached.
spinner.startLoading()
MapboxNavigationApp.getObserver(CarRouteRequest::class).request(placeRecord) {
  spinner.stopLoading()
}
```